### ISSUE TITLE
PyGMO pip install "no matching distribution" on Python 2.7.15

### ISSUE THREAD

**Issue by andrewbates09** [esa/pagmo2#237]
When attempting to install pygmo on CentOS (RHEL) Linux, I see the following mixed results, depending on the version of python used.  In both cases, pre-installed python modules are identical.

### Python 2.7.15 -- failed
* Python built with GCC 7.3.0
```
#> pip search pygmo
pygmo-plugins-nonfree (0.5)
pygmo (2.9)

#> pip install --user pygmo
Could not find a version that satisfies the requirement pygmo (from version: )
No matching distribution found for pygmo
```

### Python 3.7.1 -- success
* Python built with GCC 7.3.0
```
#> pip3 search pygmo
pygmo-plugins-nonfree (0.5)
pygmo (2.9)

#> pip3 install --user pygmo
Collecting ...
Downloading ...
Requirement(s) already satisfied ...
Successfully installed pygmo-2.9
```

**Comment by Bacoknight** [comment]
I also had this issue, but was able to install it using the .whl file:

https://pypi.org/project/pygmo/#files

**Comment by bluescarni** [comment]
I am not an expert on the topic, but as far as I remember there are different ways one can build Python 2 (more than Python 3). Indeed, if you look at the filenames of the ``whl`` files, you can see that the Python 2 package name contains the string ``cp27mu``, which, I believe, stands for CPython 2.7 (``cp27``), built with pymalloc (``m``) and wide unicode (``u``, not sure what it means exactly, as far as I know the ``u`` tag is not there in Python 3 builds).

I think that ``pip`` will refuse to install whl packages which have been built against a Python version configured differently from the one ``pip`` is using. See a bit more details here as well:

https://github.com/pypa/manylinux

**Comment by andrewbates09** [comment]
@Bacoknight thanks for your response.  I gave this a whirl, but encountered the following error:
```
#> pip install --user pygmo-2.9-cp27-cp27mu-manylinux1_x86_64.whl
pygmo-2.9-cp27-cp27mu-manylinux1_x86_64.whl is not a supported wheel on this platform.
```
This is a good transition into @bluescarni comments about the wheel type:
> ... the Python 2 package name contains the string `cp27mu`, which, I believe, stands for CPython 2.7 (`cp27`), built with pymalloc (`m`) and wide unicode (`u`, not sure what it means exactly, as far as I know the `u` tag is not there in Python 3 builds). ...

When taking a quick inspection of a [few](https://pypi.org/project/numpy/#files) [other](https://pypi.org/project/scipy/#files) [packages](https://pypi.org/project/Cython/#files), I see that the following two options are consistently available wheel files:
* *-cp27m-manylinux1_x86_64.whl
* *-cp27mu-manylinux1_x86_64.whl

This would encourage me to think that including both options in the pagmo2 files/install would be of some benefit, if not solve this issue completely.

**Comment by bluescarni** [comment]
I don't think it should be too much of a problem. Pinging @darioizzo as he's also in the loop regarding the manylinux builds, so perhaps he can tell whether this is easy to do or not. I'd certainly be oriented towards including this extra build for the next version.

**Comment by darioizzo** [comment]
It probably enough to modify the build system as to use a different folder here:
https://github.com/esa/pagmo2/blob/master/tools/install_docker.sh#L20

I agree that we should add this as should be straight forward

**Comment by andrewbates09** [comment]
@darioizzo could the current release of pygmo 2 be built this way or is there an upcoming release in the works that this will be applied to?

**Comment by bluescarni** [comment]
@andrewbates09 I need to finish up the PR about parallel island initialization, then we will make a 2.10 release with the necessary changes for better Python 2.7 support.

**Comment by andrewbates09** [comment]
@bluescarni thanks for the update.  Do you have any ETA on such things?

I am working with some researchers who are doing multidimensional bayesian optimization work and have that as a dependency.

**Comment by bluescarni** [comment]
I haven't been able to work on pagmo as much as I wanted to in the last month. But I am confident I can get it done during the holidays.

**Comment by bluescarni** [comment]
@andrewbates09 The ``py27m`` builds are now enabled in #239 and it looks like they are working. I'll start the release process of pagmo/pygmo 2.10 soon (without the parallel init stuff, which will be in 2.11 instead).

A word of warning. While working on this, I noticed that there are no ``py27m`` binary packages for pyzmq. pyzmq is a dependency of ipyparallel, which in turn is needed to get parallel optimisation working on pygmo with Python 2.7. So if you do ``pip install ipyparallel`` on a ``py27m`` setup, ``pip`` will try to compile and install pyzmq from source, which in turn requires some installation of the zeromq library on the system.

The bottom line is, it might be a bit painful to get parallelization working on a ``py27m`` setup.

https://pypi.org/project/pyzmq/#files
